rust/src/shards/gui/layouts/group.rs

```diff
@@ -0,0 +1,151 @@
+use crate::types::Parameters;
+use crate::types::ShardsVar;
+use crate::types::Type;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::ANY_TYPES;
+use crate::types::SHARDS_OR_NONE_TYPES;
+
+lazy_static! {
+  static ref GROUP_PARAMETERS: Parameters = vec![(
+    cstr!("Contents"),
+    cstr!("The UI contents."),
+    &SHARDS_OR_NONE_TYPES[..],
+  )
+    .into(),];
+}
+
+impl Default for Group {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      contents: ShardsVar::default(),
+    }
+  }
+}
+
+impl Shard for Group {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.Group")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.Group-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.Group"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&GROUP_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => self.contents.set_param(value),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.contents.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    let exp_info = ExposedInfo {
+      exposedType: EGUI_UI_SEQ_TYPE,
+      name: self.parents.get_name(),
+      help: cstr!("The parent UI objects.").into(),
+      ..ExposedInfo::default()
+    };
+    self.requiring.push(exp_info);
+
+    Some(&self.requiring)
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    if !self.contents.is_empty() {
+      let outputType = self.contents.compose(&data)?;
+      return Ok(outputType);
+    }
+
+    Ok(data.inputType)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.parents.warmup(ctx);
+    if !self.contents.is_empty() {
+      self.contents.warmup(ctx)?;
+    }
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    if !self.contents.is_empty() {
+      self.contents.cleanup();
+    }
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if self.contents.is_empty() {
+      return Ok(*input);
+    }
+
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      let output = ui
+        .group(|ui| {
+          util::activate_ui_contents(context, input, ui, &mut self.parents, &mut self.contents)
+        })
+        .inner?;
+
+      Ok(output)
```

Comment:
should be `*input`.

Reply:
Propagation of the output of the inner shards (see comment on collapsing header and README).